builder/pipeline.cc

```diff
@@ -1264,13 +1321,41 @@ class pipeline_builder {
       all_deps.insert(v);
     }
 
+    // We actually don't need to produce a full crop, but only dimensions
+    // which (transitively) depend on the this loop's base function crop_dim.
+    std::map<var, std::vector<int>> dependent_dims;
+    // Add a dimension which will have crop_dim.
+    if (base_f) {
+      for (const func::output& o : base_f->outputs()) {
+        for (int d = 0; d < static_cast<int>(o.dims.size()); ++d) {
+          if (o.dims[d] == loop.sym()) {
+            dependent_dims[o.sym()].push_back(d);
+          }
+        }
+      }
+    }
+
+    // Propagate dependent dims starting from the outer crop.
+    for (auto i = order_.begin(); i != order_.end(); ++i) {
+      const func* f = *i;
+      if (f == base_f) continue;
+      for (const func::output& o : f->outputs()) {
+        const buffer_expr_ptr& b = o.buffer;
+        if (!inferred_bounds_[b->sym()]) continue;
+        if (all_deps.count(b->sym()) == 0) continue;
+        if (body.allocations.count(b->sym()) > 0) continue;
```

Comment:
Is this skipping allocations that are inside the loop? It's a little surprising that we need to explicitly do this, and there isn't some other mechanism that would prevent such crops from being generated. Wouldn't this be an error?